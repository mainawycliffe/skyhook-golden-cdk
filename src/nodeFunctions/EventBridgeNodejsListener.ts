import { Alarm } from 'aws-cdk-lib/aws-cloudwatch';
import {
  NodejsFunctionProps,
  BundlingOptions,
} from 'aws-cdk-lib/aws-lambda-nodejs';
import { Rule, RuleProps } from 'aws-cdk-lib/aws-events';
import { Duration } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { LambdaFunction } from 'aws-cdk-lib/aws-events-targets';
import { Queue } from 'aws-cdk-lib/aws-sqs';
import SkyhookNodejsFunction from './SkyhookNodejsFunction';

/**
 * EventBridge Nodejs Listener
 *
 * Creates a SkyhookNodejsFunction {@link SkyhookNodejsFunction} & hooks in up
 * to an Appsync Resolver.
 *
 * We recommend using the typescript type `EventBridgeHandler` from
 * `types/aws-lambda`, within your handler code.
 */
export default class EventBridgeNodejsListener extends SkyhookNodejsFunction {
  constructor(
    scope: Construct,
    id: string,
    {
      description,
      entry,
      eventBus,
      eventPattern,
      handler,
      bundling,
      timeout,
    }: EventBridgeNodejsListenerProps
  ) {
    super(scope, `${id}Lambda`, {
      entry,
      description,
      handler,
      bundling,
      timeout,
    });

    const deadLetterQueue = new Queue(scope, `${id}DeadLetterQueue`, {
      retentionPeriod: Duration.days(14),
    });

    const ruleTarget = new LambdaFunction(this, {
      deadLetterQueue,
    });

    new Rule(scope, `${id}Rule`, {
      targets: [ruleTarget],
      eventPattern,
      eventBus,
      description,
    });

    new Alarm(scope, `${id}Alarm`, {
      alarmDescription: `Event listener ${id} has failed to process an event.`,
      evaluationPeriods: 1,
      metric: deadLetterQueue.metricNumberOfMessagesReceived(),
      threshold: 1,
    });
  }
}

interface EventBridgeNodejsListenerProps {
  /** Description for both the lambda and the EventBridge rule */
  description?: string;
  entry: NodejsFunctionProps['entry'];
  eventBus?: RuleProps['eventBus'];
  eventPattern: RuleProps['eventPattern'];
  handler?: NodejsFunctionProps['handler'];
  bundling?: BundlingOptions;
  timeout?: NodejsFunctionProps['timeout'];
}
