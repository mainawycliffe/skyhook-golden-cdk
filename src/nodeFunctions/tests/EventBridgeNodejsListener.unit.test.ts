import "@aws-cdk/assert/jest";
import { Stack } from "aws-cdk-lib/core";
import { SynthUtils } from "@aws-cdk/assert";
import EventBridgeNodejsListener from "../EventBridgeNodejsListener";

const testStack = new Stack();
new EventBridgeNodejsListener(testStack, "testFunction", {
  entry: "dummyEntry",
  eventPattern: { detailType: ["EventDetailTypeA"] },
});

it("creates a Lambda Function", () => {
  expect(testStack).toHaveResourceLike("AWS::Lambda::Function");
});

it("creates a rule with the provided event pattern", () => {
  // https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-events-rule.html#cfn-events-rule-eventpattern
  expect(testStack).toHaveResourceLike("AWS::Events::Rule", {
    EventPattern: {
      "detail-type": ["EventDetailTypeA"],
    },
  });
});

it("creates a dead letter queue for failures", () => {
  // https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-sqs-queues.html
  expect(testStack).toHaveResourceLike("AWS::SQS::Queue", {
    MessageRetentionPeriod: 1209600, // 14 days (maximum)
  });
});

it("sets the event target as the lambda, with a dead letter queue", () => {
  // Note we have to put hardcoded random IDs from the snapshot below as there is no way to set these
  // https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-events-rule-target.html
  expect(testStack).toHaveResourceLike("AWS::Events::Rule", {
    Targets: [
      {
        Arn: {
          "Fn::GetAtt": ["testFunctionLambda78269C17", "Arn"],
        },
        DeadLetterConfig: {
          Arn: {
            "Fn::GetAtt": ["testFunctionDeadLetterQueue126C440D", "Arn"],
          },
        },
        Id: "Target0",
      },
    ],
  });
});

it("creates an alarm on the dead letter queue", () => {
  expect(testStack).toHaveResourceLike("AWS::CloudWatch::Alarm", {
    MetricName: "NumberOfMessagesReceived",
    Namespace: "AWS/SQS",
  });
});

it("creates resources matching the snapshot", () => {
  const cfn = SynthUtils.toCloudFormation(testStack);
  expect(cfn).toMatchSnapshot();
});
