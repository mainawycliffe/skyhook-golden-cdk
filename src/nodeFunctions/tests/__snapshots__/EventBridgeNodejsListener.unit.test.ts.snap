// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates resources matching the snapshot 1`] = `
Object {
  "Resources": Object {
    "testFunctionAlarm403B59EF": Object {
      "Properties": Object {
        "AlarmDescription": "Event listener testFunction has failed to process an event.",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": Array [
          Object {
            "Name": "QueueName",
            "Value": Object {
              "Fn::GetAtt": Array [
                "testFunctionDeadLetterQueue126C440D",
                "QueueName",
              ],
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "NumberOfMessagesReceived",
        "Namespace": "AWS/SQS",
        "Period": 300,
        "Statistic": "Sum",
        "Threshold": 1,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "testFunctionDeadLetterQueue126C440D": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "MessageRetentionPeriod": 1209600,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "testFunctionDeadLetterQueuePolicyB029CBEC": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sqs:SendMessage",
              "Condition": Object {
                "ArnEquals": Object {
                  "aws:SourceArn": Object {
                    "Fn::GetAtt": Array [
                      "testFunctionRule4EC511D3",
                      "Arn",
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Service": "events.amazonaws.com",
              },
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testFunctionDeadLetterQueue126C440D",
                  "Arn",
                ],
              },
              "Sid": "AllowEventRuletestFunctionRule",
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": Array [
          Object {
            "Ref": "testFunctionDeadLetterQueue126C440D",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "testFunctionLambda78269C17": Object {
      "DependsOn": Array [
        "testFunctionLambdaServiceRoleF9778236",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": "cdk-lambda-nodejs-mock-fake-bucket",
          "S3Key": "cdk-lambda-nodejs-mock-fake-lambda.ts",
        },
        "MemorySize": 1024,
        "Role": Object {
          "Fn::GetAtt": Array [
            "testFunctionLambdaServiceRoleF9778236",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "testFunctionLambdaServiceRoleF9778236": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "testFunctionRule4EC511D3": Object {
      "Properties": Object {
        "EventPattern": Object {
          "detail-type": Array [
            "EventDetailTypeA",
          ],
        },
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "testFunctionLambda78269C17",
                "Arn",
              ],
            },
            "DeadLetterConfig": Object {
              "Arn": Object {
                "Fn::GetAtt": Array [
                  "testFunctionDeadLetterQueue126C440D",
                  "Arn",
                ],
              },
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "testFunctionRuleAllowEventRuletestFunctionLambdaD644A83C": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "testFunctionLambda78269C17",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "testFunctionRule4EC511D3",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
}
`;
